sim/mage/arcane_missiles.go

```diff
@@ -0,0 +1,86 @@
+package mage
+
+import (
+	"time"
+
+	"github.com/wowsims/cata/sim/core"
+)
+
+func (mage *Mage) registerArcaneMissilesSpell() {
+
+	mage.ArcaneMissilesTickSpell = mage.GetOrRegisterSpell(core.SpellConfig{
+		ActionID:       core.ActionID{SpellID: 7268},
+		SpellSchool:    core.SpellSchoolArcane,
+		ProcMask:       core.ProcMaskSpellDamage | core.ProcMaskNotInSpellbook,
+		Flags:          SpellFlagMage,
+		ClassSpellMask: MageSpellArcaneMissilesTick,
+		MissileSpeed:   20,
+
+		DamageMultiplier: 1,
+		CritMultiplier:   mage.DefaultSpellCritMultiplier(),
+		ThreatMultiplier: 1,
+		BonusCoefficient: 0.278,
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			damage := 0.432 * mage.ScalingBaseDamage
+			result := spell.CalcDamage(sim, target, damage, spell.OutcomeMagicHitAndCrit)
+			//spell.WaitTravelTime(sim, func(sim *core.Simulation) {
+			spell.DealDamage(sim, result)
+			//})
+		},
+	})
+
+	var numTicks int32 = 3
+	mage.ArcaneMissiles = mage.RegisterSpell(core.SpellConfig{
+		ActionID:       core.ActionID{SpellID: 7268},
+		SpellSchool:    core.SpellSchoolArcane,
+		ProcMask:       core.ProcMaskSpellDamage,
+		Flags:          SpellFlagMage | core.SpellFlagChanneled | core.SpellFlagAPL,
+		ClassSpellMask: MageSpellArcaneMissilesCast,
+		Cast: core.CastConfig{
+			DefaultCast: core.Cast{
+				GCD: core.GCDDefault,
+			},
+		},
+		ExtraCastCondition: func(sim *core.Simulation, target *core.Unit) bool {
+			return mage.ArcaneMissilesProcAura.IsActive()
+		},
+		Dot: core.DotConfig{
+			Aura: core.Aura{
+				Label: "ArcaneMissiles",
+				OnGain: func(aura *core.Aura, sim *core.Simulation) {
+					// Force cast a missile right away
+					mage.ArcaneMissilesTickSpell.Cast(sim, mage.CurrentTarget)
+					mage.ArcaneMissilesProcAura.Deactivate(sim)
+				},
+				OnExpire: func(aura *core.Aura, sim *core.Simulation) {
+					if mage.ArcaneMissilesProcAura.IsActive() {
+						mage.ArcaneMissilesProcAura.Deactivate(sim)
+					}
+
+					// TODO: This check is necessary to ensure the final tick occurs before
+					// Arcane Blast stacks are dropped. To fix this, ticks need to reliably
+					// occur before aura expirations.
+					dot := mage.ArcaneMissiles.Dot(aura.Unit)
+					if dot.TickCount < dot.NumberOfTicks {
+						dot.TickCount++
+						dot.TickOnce(sim)
+					}
+					mage.ArcaneBlastAura.Deactivate(sim)
+				},
+			},
+			NumberOfTicks:       numTicks - 1, // subtracting 1 due to autocasting one OnGain
+			TickLength:          time.Millisecond * 700,
+			AffectedByCastSpeed: true,
```

Comment:
I'm not sure if arcane missiles work like other hasted dots. Is this supposed to keep channeling for the original duration and just send more missiles if you have more haste? Thats whats gonna happen with the current dot system.

Reply:
It should be a static number of ticks, standard 3, up to 5 based on talents. in testing this implementation, adding haste only shortens the duration and does not add ticks.

Reply:
Interesting... I would've expect it to work like any other dot we have which would add more ticks to fit the duration

Reply:
Actually you're right, it just shows up differently than I expected in the timeline. Logs do show rapid casting, but timeline only shows 5 "ticks" causing a cast.